Math/NumberTheory/Primes/Factorisation/QuadraticSieve.hs

```diff
@@ -14,173 +14,222 @@ import Data.Semigroup
-
--- | Given an odd positive composite Integer @n@ and Int parameters @b@ and @t@,
--- the Quadratic Sieve attempts to output @factor@, a factor of @n@. The
+import Math.NumberTheory.Roots
+-- | Given an odd positive composite Integer @n@ and Int parameters @b@, @t@ and
+-- @k@, the Quadratic Sieve attempts to output @factor@, a factor of @n@. The
 -- parameter @b@ controls the size of the factor base. This consists of all
 -- the relevant primes which are less than or equal to @b@. The parameter @t@
--- controls the length of the sieving interval.
-quadraticSieve :: Integer -> Int -> Int -> Integer
-quadraticSieve n b t = findFactor n $ findSquares n b t
+-- controls the length of the sieving interval. The parameter @k@ controls the
+-- maximum number of blocks for the sieve to go through (@2 ^ (k - 1)@).
+quadraticSieve :: Integer -> Int -> Int -> Int -> Integer
+quadraticSieve n t m k = findFactor n $ findSquares n t m k
 
 findFactor :: Integer -> [(Integer, Integer)] -> Integer
-findFactor _ [] = error "Quadratic Sieve failed."
+findFactor _ [] = error "Parameters are not large enough."
 findFactor n ((x, y) : otherSquares)
-  | factor /= 1 && factor /= n = factor
-  | otherwise                  = findFactor n otherSquares
+  | (x * x - y * y) `mod` n /= 0 = error ("Algorithm incorrect." ++ show (x, y))
+  | factor /= 1 && factor /= n   = factor
+  | otherwise                    = findFactor n otherSquares
   where
     factor = gcd (x - y) n
 
 -- This routine outputs an infinite list of tuples @(x, y)@ such that
 -- @x ^ 2 - y ^ 2 `mod` n = 0@. A factorisation can be infered from this data
 -- in at least a half of the cases.
-findSquares :: Integer -> Int -> Int -> [(Integer, Integer)]
-findSquares n b t = runST $ do
+findSquares :: Integer -> Int -> Int -> Int -> [(Integer, Integer)]
+-- k indicates maximum number of sieve blocks to go thorugh 2 ^ (k - 1)
+findSquares n t m k = runST $ do
   let
-    factorBase = [nextPrime 2..precPrime b]
-    squareRoot = integerSquareRoot n
-    mappingFunction j = j * j - n
-    startingPoint = squareRoot - intToInteger t `div` 2
+    factorBase = [nextPrime 2..precPrime t]
+    -- Make sure rootOfA is an Int
+    rootOfA = ((2 * fromInteger n) ** (1 / (4 * fromIntegral k))) / (fromIntegral m ** (1 / (2 * fromIntegral k))) :: Double
+    listOfFactors = trace ("Root of a: " ++ show rootOfA) $ generatePrimes n rootOfA k
+    decompositionOfA = zip (map (fromJust . toPrimeIntegral) listOfFactors) (repeat 2)
+    a = foldr (\(p, i) acc -> acc * unPrime p ^ i) 1 decompositionOfA
+    valuesOfB = trace ("Ratio: " ++ show ((fromInteger (a * fromIntegral m) :: Double) / (fromInteger (integerSquareRoot (2*n))))) $ filter (<= a `div` 2) $ sqrtsModFactorisation n decompositionOfA
+    valuesOfC = map (\x -> (x * x - n) `div` a) valuesOfB
+    mappingFunctions = zipWith (curry (\ (b, c) x -> a * x * x + 2 * b * x + c)) valuesOfB valuesOfC
```

Comment:
Probably does not make much sense to use `curry` - just pattern-match on a nested tuple `((b, c), x)`.